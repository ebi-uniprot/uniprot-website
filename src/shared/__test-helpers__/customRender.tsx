/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
