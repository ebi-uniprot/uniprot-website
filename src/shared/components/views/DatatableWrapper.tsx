import {
  useState,
  useRef,
  useEffect,
  useLayoutEffect,
  HTMLAttributes,
} from 'react';
import { Button } from 'franklin-sites';
import { useParams } from 'react-router-dom';
import cn from 'classnames';

import useCustomElement from '../../hooks/useCustomElement';

import { sendGtagEventFeatureDataTableViewClick } from '../../utils/gtagEvents';

import './styles/datatable-wrapper.scss';

type Props = {
  alwaysExpanded?: boolean;
} & HTMLAttributes<HTMLDivElement>;

const DatatableWrapper = ({
  alwaysExpanded,
  children,
  className,
  ...props
}: Props) => {
  const [showButton, setShowButton] = useState(!alwaysExpanded);
  const [expandTable, setExpandTable] = useState(false);

  const tableRef = useRef<HTMLElement>(null);
  const firstRenderRef = useRef(true);

  const params = useParams<{ accession?: string }>();

  const datatableElement = useCustomElement(
    /* istanbul ignore next */
    () =>
      import(
        /* webpackChunkName: "protvista-datatable" */ 'protvista-datatable'
      ),
    'protvista-datatable'
  );

  // On expand/collapse change
  useEffect(() => {
    // except on first render or when always expanded
    if (!alwaysExpanded && !firstRenderRef.current) {
      // Scroll table back into view when collapsing
      if (!expandTable) {
        tableRef.current?.parentElement?.scrollIntoView({
          behavior: 'smooth',
          block: 'center',
        });
      }
      sendGtagEventFeatureDataTableViewClick(
        params.accession || '',
        expandTable ? 'expanded' : 'collapsed'
      );
    }
    // If first render was previous render, then it's not the first anymore...
    if (firstRenderRef.current) {
      firstRenderRef.current = false;
    }
  }, [alwaysExpanded, expandTable, params.accession]);

  // eslint-disable-next-line consistent-return
  useLayoutEffect(() => {
    if (!alwaysExpanded && datatableElement.defined && tableRef.current) {
      const mo = new MutationObserver(() => {
        setShowButton(
          tableRef.current?.shadowRoot?.firstElementChild?.scrollHeight !==
            tableRef.current?.shadowRoot?.firstElementChild?.clientHeight
        );
      });
      mo.observe(tableRef.current, {
        childList: true,
        // Allows to update when classes are applied to hide some rows
        attributes: true,
        attributeFilter: ['class'],
        subtree: true,
      });
      return () => mo.disconnect();
    }
  }, [alwaysExpanded, datatableElement.defined]);

  // if (datatableElement.defined) {
  //   return (
  //     <div className={cn('datatable-wrapper', className)} {...props}>
  //       <datatableElement.name
  //         ref={tableRef}
  //         filter-scroll
  //         expandtable={alwaysExpanded || expandTable ? '' : undefined}
  //       >
  //         {children}
  //       </datatableElement.name>
  //       {(showButton || expandTable) && (
  //         <Button
  //           variant="primary"
  //           onClick={() => setExpandTable((current) => !current)}
  //           className="toggle-button"
  //         >
  //           {expandTable ? 'Collapse' : 'Expand'} table
  //         </Button>
  //       )}
  //     </div>
  //   );
  // }
  return (
    <div className={cn('datatable-wrapper', className)} {...props}>
      {children}
    </div>
  );

  return null;
};

export default DatatableWrapper;
