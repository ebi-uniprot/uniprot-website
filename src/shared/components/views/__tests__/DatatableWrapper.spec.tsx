import { fireEvent, screen } from '@testing-library/react';

import customRender from '../../../__test-helpers__/customRender';

import DatatableWrapper from '../DatatableWrapper';

describe('DatatableWrapper component', () => {
  it('should render with a working toggle button', async () => {
    const { asFragment } = customRender(
      <DatatableWrapper>
        <table />
      </DatatableWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
    const button = screen.getByRole('button', { name: /Expand/ });
    expect(button).toBeInTheDocument();
    fireEvent.click(button);
    expect(
      await screen.findByRole('button', { name: /Collapse/ })
    ).toBeInTheDocument();
  });

  it('should not render a toggle button if overriden', async () => {
    const { asFragment } = customRender(
      <DatatableWrapper alwaysExpanded>
        <table />
      </DatatableWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });
});
