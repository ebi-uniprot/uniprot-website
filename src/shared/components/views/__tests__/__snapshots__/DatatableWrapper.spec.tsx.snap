// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatatableWrapper component should not render a toggle button if overriden 1`] = `
<DocumentFragment>
  <div
    class="datatable-wrapper"
  >
    <protvista-datatable
      expandtable=""
      filter-scroll="true"
    >
      <table />
    </protvista-datatable>
  </div>
</DocumentFragment>
`;

exports[`DatatableWrapper component should render with a working toggle button 1`] = `
<DocumentFragment>
  <div
    class="datatable-wrapper"
  >
    <protvista-datatable
      filter-scroll="true"
    >
      <table />
    </protvista-datatable>
    <button
      class="button primary toggle-button"
      type="button"
    >
      Expand table
    </button>
  </div>
</DocumentFragment>
`;
